### PR TITLE
test: use puppeteer to test all examples

### PR DESCRIPTION
**Pull request by peppsac**
This commit replaces the itowns-testing.js hack with puppeteer (https://github.com/GoogleChrome/puppeteer).

Advantages:
  - a real browser (Chrome) instead of node + mocks for missing features
  - WebGL capable
  - Can take screenshots and save them (not used on Travis yet)
  - Debuggable using Chrome Dev tools (Chrome instance used by puppeteer can be headless or not)

Each example now have at least one test to verify we can open it and that it
reaches the "mainloop has finished running" state.

I also implemented a couple tests for other examples (see 3dtiles and globe).

Adding test is easy, and we could add more to exercise more parts of itowns.

Last but not least: with this change the separation of examples in 2 files (html + js) is
not needed anymore.

Environment variables supported:
  - `USE_DEV_SERVER=1`: do not start a new itowns server (npm start). Useful if you already have a itowns server running
  - `SCREENSHOT_FOLDER=/tmp`: if specified, tests will save a screenshot in this folder (must exist)
 - `DEBUG=1`: launch Chrome in non-headless mode

### CODE REVIEW

**Comment by gchoqueux** [comment]
@peppsac  The PR's subject  seems good. But for such a choice, it seems more appropriate to open an issue in advance. (to desist together) 
Thanks

**Comment by zarov** [comment]
Good thing to see the merge of html + js files, we're going to clearer sample like threejs. To reduce this PR, this could be done in another one, or at least in another commit.

Otherwise I couldn't install puppeteer, with the error a lot of people seems to encounter [(`ERROR: Failed to download Chromium r549031! Set "PUPPETEER_SKIP_CHROMIUM_DOWNLOAD"`)](https://github.com/GoogleChrome/puppeteer/issues/2173). I know this might be because I am behing a proxy/firewall, but this is a no for me since I think users outside the iTowns team could be in the same situation, and I don't expect them to spend too much time configuring everything to make puppeteer work.

**Comment by peppsac** [comment]
@gchoqueux :
> @peppsac The PR's subject seems good. But for such a choice, it seems more appropriate to open an issue in advance. (to desist together)

I prefer to first implement something to see if it's worth it. So I'll probably keep doing like this.
This doesn't mean that I force my choices on you: this still has to be approved before being merged.

@zarov :
> this is a no for me since I think users outside the iTowns team could be in the same situation

`puppeteer` is added as a dev dependency, so this issue will only affect iTowns developers, not regular itowns users.

**Comment by gchoqueux** [comment]
As this PR proposes a technological choice, this choice can be revoked. To avoid wasting time in development, reviewing and testing (and not a story of forcing), it's better for the community to make this choice in advance. 
And don't confuse your preferences with those of the community

**Comment by peppsac** [comment]
> As this PR proposes a technological choice, this choice can be revoked.

Yep. Feel free to expose your arguments against this change so we can start a discussion.

> To avoid wasting time in development, reviewing and testing (and not a story of forcing), it's better for the community to make this choice in advance

That's your call. 
But I think I can spend my time like I want :)

>  And don't confuse your preferences with those of the community

You're misunderstanding what I wrote.

I wrote "**I** prefer to write code and then ask a review". It's quite different from "everyone must write code before discussing".

**Comment by gchoqueux** [comment]
For important changes, it is advisable to notify the community in advance. [see](https://github.com/iTowns/itowns/blob/master/CONTRIBUTING.md#opening-a-pull-request)

It's better to validate in advance to avoid the time lost by the reviewers if finally the technological choice isn't validated

Could you add installation instructions with proxy (or link) for puppeteer

Could you split commit : clean examples and puppeteer

**Comment by peppsac** [comment]
> For important changes, it is advisable to notify the community in advance. see

This document is a guide for new contributors.
Because closing a PR for a newcomer might be rude, this document suggests to open an issue.
But I'm not a newcomer regarding itowns, so I won't be offended if we decide that this PR isn't worth it.

> It's better to validate in advance to avoid the time lost by the reviewers if finally the technological choice isn't validated

Discussing in an issue or in a PR isn't "time lost".
And at least with a PR you can actually see the changes and test them, rather than discussing vague ideas that might not even work.

> Could you add installation instructions with proxy (or link) for puppeteer

I'll amend the README with the link @zarov posted above (https://github.com/GoogleChrome/puppeteer/issues/2173#issuecomment-372727628)

> Could you split commit : clean examples and puppeteer

I'll do this before merging.
Though the first one won't be "clean example" but "test: remove itown- testing" (because it's the removal of itowns-testing.js that allows to merge the .js and .html)

**Comment by gchoqueux** [comment]
This document is a guide for all contributors

I'm talking about the loss of time in reviewing code, however there's no waste of time on the discussion to choice a technological (especially in a dedicated issue).

(https://github.com/GoogleChrome/puppeteer/issues/2173#issuecomment-372727628) it needs to be applied to every chromium update, it's not a viable solution.

The split is to simplify reviewing, so it's best to do it before the merge

**Comment by peppsac** [comment]
>This document is a guide for all contributors
> I'm talking about the loss of time in reviewing code, however there's no waste of time on the discussion to choice a technological (especially in a dedicated issue).

@gchoqueux ok I'm done: I won't answer any of your comments not directly related to the code / design of this PR.

**Comment by nosy-b** [comment]
@peppsac Thanks for the great work, puppeteer looks promising :smiley: 

- But yes you could have started an issue on that subject 
- The PR should show that it is _WIP_, or a _Suggestion for enhancement_
- Too bad that we need some special tweeking to make it work. -> Not yet ready for itowns

**Comment by peppsac** [comment]
> But yes you could have started an issue on that subject

See my previous answers.

> The PR should show that it is WIP, 

This is not a WIP PR since it's fully working.

> or a Suggestion for enhancement

Done

> Too bad that we need some special tweeking to make it work. -> Not yet ready for itowns

So, between having no test at all for itowns and having to do a one time setup on your machine you choose the no-test option?

I pushed an alternative: I removed puppeteer from our dev-dependency and install it manually on travis.

**Comment by peppsac** [comment]
@zarov : did you try with the documented environment variables https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#environment-variables ? 

Or another simpler way to fix it:
- install Chrome (recent enough, Chrome beta works fine: https://www.google.fr/chrome/browser/beta.html)
- install puppeteer without downloading Chrome: `PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=1 npm install puppeteer`
- then use the env var `CHROME` to tell itowns/mocha/puppeteer what Chrome app it should use:
`CHROME=/opt/google/chrome-beta/chrome npm run test-examples`

I don't think it's that much complicated, is it?

And if `puppeteer` is not acceptable, fine but what solution do you propose that you can commit to work on in a near future?

**Comment by zarov** [comment]
Fair enough, it works following these instructions, I didn't try to fix the problem before. Those need to amended to the readme.

Love the screenshot feature ! However, it seems to overwrite a previous screenshot for an example that have multiple tests (see 3dtiles).
I can also see that some tests should be failing, as they are not correctly rendered: it is the case for `3dtiles` (white screenshot), `panorama` (black), `gpx` (no gpx trace visible). I also have a failing test for `positionGlobe`, but I think you are working on it ?

**Comment by peppsac** [comment]
> Love the screenshot feature ! However, it seems to overwrite a previous screenshot for an example that have multiple tests (see 3dtiles).

Yes, I have a not-yet-pushed modification that allows to name screenshots to avoid this problem

> I can also see that some tests should be failing, as they are not correctly rendered: it is the case for 3dtiles (white screenshot), panorama (black),

Not sure whatr's wrong here. Can you run with `DEBUG=1` so you can see what's going on in the debugger?

`positionGlobe` test is expected to fail, until PR #748  is merged.

File: test/examples/boostrap.js
Comment (zarov) [inline test/examples/boostrap.js]: I see no problem with moving this to its own file, like in `test/examples/server.js` maybe ?

Comment (peppsac) [reply test/examples/boostrap.js]: Yeah, I did it this way to only have test case in this folder. But this can be changed.

**Comment by peppsac** [comment]
Rebased.

I also squashed the commits and move the 2 `View` fixes to PR #749.
And pushed a fix to make sure the loading screen is hidden when taking the screenshots.

Now waiting for comments on this PR (ping @iTowns/reviewers).

**Comment by peppsac** [comment]
@gchoqueux:
> Could you split commit : clean examples and puppeteer

is done (thanks to #750)

File: src/Core/View.js
Comment (tbroyer) [inline src/Core/View.js]: I suppose this rewrite was to avoid the spread operator. How about using this one-liner instead then?
```js
Array.prototype.push.apply(result, layer.pickObjectsAt(this, mouse));
```
(see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/push#Merging_two_arrays)

File: test/examples/boostrap.js
Comment (tbroyer) [inline test/examples/boostrap.js]: Should this be disabled for the whole `test/` folder?

File: test/examples/boostrap.js
Comment (tbroyer) [inline test/examples/boostrap.js]: Any reason you need to fork? Couldn't you run the server in the same process? (and btw use the `listening` event for `waitServerReady`)

You could then use:
* the `listening` event for `waitServerReady` (sure you'd have to rework the way `USE_DEV_SERVER` is handled, but that should be pretty easy as first glance)
* port 0 (i.e. let NodeJS find an available port, and retrieve it with `server.address().port`) instead of `findFreeTcpPort`
* https://www.npmjs.com/package/serve-static as the request listener

File: test/examples/boostrap.js
Comment (tbroyer) [inline test/examples/boostrap.js]: call `resolve(false)` and return `Promise.reject()` like below?

File: test/examples/boostrap.js
Comment (tbroyer) [inline test/examples/boostrap.js]: nit: authorize `_` in addition to `\w`?

File: test/examples/boostrap.js
Comment (tbroyer) [inline test/examples/boostrap.js]: Could you document why those flags are needed?

File: test/examples/.eslintrc.js
Comment (tbroyer) [inline test/examples/.eslintrc.js:13]: Add Mocha? https://eslint.org/docs/user-guide/configuring#specifying-environments (IIUC, that'll save you from manually declaring its globals)

**Review by tbroyer — COMMENTED** [review]
I wholeheartedly support this initiative. I'll let others eventually approve the PR though.

File: src/Core/View.js
Comment (peppsac) [reply src/Core/View.js]: The problem was `javascript maximum call stack size exceeded` because the array might be very large.
Your suggestion unfortunately won't work either (`Do not use this method if the second array`) but I'll add a comment to explain why we need to do this.

File: test/examples/boostrap.js
Comment (peppsac) [reply test/examples/boostrap.js]: It's only used here, because `puppeteer` must be manually installed (the other tes files don't directly depend on `puppeteer` so they don't need the `eslint-disable` trick).
So I'd say we can keep it here, WDYT?

File: test/examples/boostrap.js
Comment (peppsac) [reply test/examples/boostrap.js]: Nope, the fork was the historical reasons, I removed it.

I've reworked this part a bit, but I'd rather not add another dependency to simply serve static files.

File: test/examples/boostrap.js
Comment (peppsac) [reply test/examples/boostrap.js]: Yep, done.

File: test/examples/.eslintrc.js
Comment (peppsac) [reply test/examples/.eslintrc.js:13]: Ah good idea, thanks

**Comment by peppsac** [comment]
Thanks for the review @tbroyer, I pushed a commit addressing your comments.

File: test/examples/boostrap.js
Comment (tbroyer) [reply test/examples/boostrap.js]: Ah OK; yeah, keep it that way then.

File: test/examples/boostrap.js
Comment (tbroyer) [inline test/examples/boostrap.js]: IIRC, documentation says you should wait for the `listening` event to fire.

How about some more refactoring?
* let `startStaticFileServer()` assign `global.itownsServer`
* also let it listen for the `listening` event and assign `global.itownsPort` there (and also do the `console.log`); and return a promise that resolves when the event is dispatched (meaning the server accepts requests, replacing the need for `_waitServerReady` in this case)
* have `waitServerReady()` only be called when `USE_DEV_SERVER` is used, `await`ing `startStaticFileServer` otherwise

```js
if (!process.env.USE_DEV_SERVER) {
  await startStaticFileServer();
} else {
  global.itownsPort = process.env.USE_DEV_SERVER;
  await waitServerReady(global.itownsPort);
}
```

Alternatively, have `createStaticFileServer`'s returned promise resolve to the port it's listening on, and `waitServerReady` resolve to the passed in port; so you can assign `global.itownsPort` at a single location:
```js
let server;
if (!process.env.USE_DEV_SERVER) {
  server = startStaticFileServer();
} else {
  server = waitServerReady(process.env.USE_DEV_SERVER);
}
// wait for the server to be ready
global.itownsPort = await server;

File: test/examples/boostrap.js
Comment (tbroyer) [inline test/examples/boostrap.js]: Remove initialization here, value is always overwritten.

File: test/examples/boostrap.js
Comment (tbroyer) [inline test/examples/boostrap.js]: Make the hook async?

```js
after(done => {
    …
    global.itownsServer.close(done);
```

File: test/examples/boostrap.js
Comment (tbroyer) [inline test/examples/boostrap.js]: That will strip `_` from the input.

```js
replace(/[^\w_]/g, '_')
```

**Review by tbroyer — DISMISSED** [review]

**Comment by peppsac** [comment]
@tbroyer thanks for your feedback, should be all fixed now.

**Comment by tbroyer** [comment]
Did I miss something or you forgot to assign `global.itownsServer`?

Otherwise SGTM

**Comment by peppsac** [comment]
> Did I miss something or you forgot to assign global.itownsServer?

Fixed, thanks!

**Comment by peppsac** [comment]
> As said above, I'll let someone else "formally" approve those changes. SGTM otherwise.

Thanks @tbroyer. Any reviewer willing to "formally" review this PR then?

File: README.md
Comment (zarov) [inline README.md:72]: It can also be `PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=1 npm install` (without the specified package) to respect the current package-lock and all.

**Review by zarov — APPROVED** [review]
Good to me too, I had problems this morning with the third test of `3dtiles`, with timeout, but it seems okay now (test done in 13s).

File: package.json
Comment (autra) [inline package.json]: these 2 dependencies should be dev dependencies.

**Review by autra — CHANGES_REQUESTED** [review]

File: package.json
Comment (peppsac) [reply package.json]: `http-server` isn't needed so I'll remove it.

And `puppeteer` won't be installed by default (even in dev dependencies) to avoid download failures.

Fixing that.

**Review by autra — APPROVED** [review]

**Comment by autra** [comment]
Thanks!

**Comment by peppsac** [comment]
Rebased, squashed + added a test for the latest added example (from #726).

I'll merge this PR soon.

**Comment by zarov** [comment]
A quick update: we have the final word on our previous problem with the proxy. It seems that in Ubuntu, when setting the `https_proxy` env variable, it automatically appends the `https` protocol to it, even if our proxy is not available in `https` and we want to access it through `http`. The solution is to export in `.profile`/`.bashrc`/`whatever` the correct variable: `export https_proxy="http://my.pro.xy"`.

So maybe now we could pass puppeteer as a dev dependency ?

**Pull request merged**
